File: src/python/philosopher_eat_for_turn.py
import enum
import time
import random
import threading

class StatesPhilosopher(enum.Enum):
    '''
    Representa los posibles estados de los filosofos.
    '''
    thinking = 1
    hungry = 2
    eating = 3
    finished_eat = 4

class Philosopher:

    def __init__(self, count_philosopher=5):
        self.count_philosopher = count_philosopher
        self.mutex = threading.Semaphore(1)
        self.state_philosopher = [ StatesPhilosopher.thinking 
                                    for _ in range(self.count_philosopher) ]
        self.start_threads()

    def eat(self, id_philosopher):
        '''
        Funcion que va ejecutar cada filosofo o hilo.
        '''
        while True:

            if random.randint(0, 2) == 0:
                self.mutex.acquire()

                self.state_philosopher[id_philosopher] = StatesPhilosopher.hungry
                print(f'[@] Filosofo {id_philosopher} hambiento.')
                
                self.state_philosopher[id_philosopher] = StatesPhilosopher.eating
                print(f'[+] Filosofo {id_philosopher} comiendo.')
                time.sleep(random.randint(3, 5))
                self.state_philosopher[id_philosopher] = StatesPhilosopher.finished_eat
                print(f'[!] Filosofo {id_philosopher} termino de comer.')
                
                self.mutex.release()
            else:
                self.state_philosopher[id_philosopher] = StatesPhilosopher.thinking
                print(f'[*] Filosofo {id_philosopher} pensando.')
                time.sleep(random.randint(2, 5))

            time.sleep(1)

    def start_threads(self):
        '''
        Comienza a correr los hilos correspondiente 
        a cada filosofo.
        '''
        threads = [ threading.Thread(target=self.eat, args=(id,)) 
                        for id in range(self.count_philosopher) ]

        for thread in threads:
            thread.start()

        for thread in threads:
            thread.join()

def main():
    
    Philosopher()

if __name__ == '__main__':
    main()

File: src/c#/philosopher_eat_for_turn.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;

namespace Philosophers
{
    public enum StatesPhilosopher
    {
        thinking = 0,
        hungry = 1,
        eating = 2,
        finished_eat = 3
    }

    public class Philosopher
    {
        private Semaphore block;
        private Thread[] philosopherThread;
        private StatesPhilosopher[] statesPhilosopher;

        public int CountPhilosopher { get; private set; }

        public Philosopher(int countPhilosopher = 5)
        {
            block = new Semaphore(1, 1);
            philosopherThread = new Thread[countPhilosopher];
            statesPhilosopher = new StatesPhilosopher[countPhilosopher];
            CountPhilosopher = countPhilosopher;

            for (int i = 0; i < statesPhilosopher.Length; i++)
            {
                statesPhilosopher[i] = StatesPhilosopher.thinking;
            }
        }

        private void Eat(int id_philosopher)
        {
            Random random = new Random();

            while (true)
            {
                if (random.Next(0, 2) == 0)
                {
                    block.WaitOne();

                    statesPhilosopher[id_philosopher] = StatesPhilosopher.hungry;
                    Console.WriteLine($"[@] Filosofo {id_philosopher} hambriento.");

                    statesPhilosopher[id_philosopher] = StatesPhilosopher.eating;
                    Console.WriteLine($"[+] Filosofo {id_philosopher} comiendo.");
                    Thread.Sleep(TimeSpan.FromSeconds(random.Next(3, 5)));
                    statesPhilosopher[id_philosopher] = StatesPhilosopher.finished_eat;
                    Console.WriteLine($"[!] Filosofo {id_philosopher} termino de comer.");

                    block.Release();
                }
                else
                {
                    statesPhilosopher[id_philosopher] = StatesPhilosopher.thinking;
                    Console.WriteLine($"[*] Filosofo {id_philosopher} pensando.");
                    Thread.Sleep(TimeSpan.FromSeconds(random.Next(2, 5)));
                }

                Thread.Sleep(TimeSpan.FromSeconds(1));
            }
        }

        public void Run()
        {
            for (int i = 0; i < philosopherThread.Length; i++)
            {
                int idx = i;
                philosopherThread[idx] = new Thread(() => Eat(idx));
                philosopherThread[idx].Start();
            }

            for (int i = 0; i < philosopherThread.Length; i++)
            {
                philosopherThread[i].Join();
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            Philosopher forTurn = new Philosopher();
            forTurn.Run();
        }
    }
}
